File: src/ReplicatedStorage/Modules/BaseModule/Relative/State/KinematicState.lua
--!strict
local Type = require(script.Parent.Parent.Parent.Parent.Type)
local Constructor = require(script.Parent.Parent.Parent.Parent.Constructor)
local State = require(script.Parent)
local Vector3D = require(script.Parent.Parent.Parent.Parent.Libraries.Vector3D)

-- Internal type
type KinematicState = Type.KinematicStateEXTENSIBLE<KinematicState,
		Type.StateEXTENSIBLE<KinematicState,
			Type.RelativeEXTENSIBLE<KinematicState,
				Type.BaseModuleEXTENSIBLE<KinematicState
	>>>>
	& Constructor.KinematicStateEXTENSIBLE<KinematicState>
	& {
	position: Type.Vector3D,
	velocity: Type.Vector3D
}

local KinematicState: KinematicState = { __type = "KinematicState" :: "KinematicState" } :: any
local KinematicStateMT = {}

function KinematicStateMT.__add(self: KinematicState, other: KinematicState): KinematicState
	self:assertTypesMatch(other)
	assert(self:sameRelativeTree(other), "KinematicState __add() Operands do not have the same relativeTree")
	-- local newSelf: KinematicState, newOther: KinematicState = self:synchronize(other)
	-- return KinematicState.new(newSelf.position + newOther.position, newSelf.velocity + newOther.velocity, newSelf:getRelativeOrNil())
	return KinematicState.new(self.position + other.position, self.velocity + other.velocity, self:getRelativeOrNil())
end

function KinematicStateMT.__sub(self: KinematicState, other: KinematicState): KinematicState
	self:assertTypesMatch(other)
	assert(self:sameRelativeTree(other), "KinematicState __sub() Operands do not have the same relativeTree")
	-- local newSelf: KinematicState, newOther: KinematicState = self:synchronize(other)
	-- return KinematicState.new(newSelf.position - newOther.position, newSelf.velocity - newOther.velocity, newSelf:getRelativeOrNil())
	return KinematicState.new(self.position - other.position, self.velocity - other.velocity, self:getRelativeOrNil())
end

function KinematicStateMT.__eq(self: KinematicState?, other: KinematicState?): boolean
	if (self ~= nil and other ~= nil) then
		self:assertTypesMatch(other)

		if (self.position == other.position and self.velocity == other.velocity) then
			if (self:hasRelative() and other:hasRelative()) then
				return KinematicStateMT.__eq(self:getRelative(), other:getRelative())
			else
				return self:hasRelative() == other:hasRelative()
			end
		else
			return false
		end
	else
		return self == nil and other == nil
	end
end

-- Constructors

--[=[
	Creates a new KinematicState instance.
]=]
function KinematicState.new(position: Type.Vector3D, velocity: Type.Vector3D, relativeTo: KinematicState?): KinematicState
	assert(
		position.X == position.X and position.Y == position.Y and position.Z == position.Z,
		"KinematicState new() position is nan (position = " .. tostring(position) .. ")"
	)
	assert(
		velocity.X == velocity.X and velocity.Y == velocity.Y and velocity.Z == velocity.Z,
		"KinematicState new() velocity is nan (velocity = " .. tostring(velocity) .. ")"
	)
	
	local self = table.clone(KinematicState) :: any
	self.position = position
	self.velocity = velocity

	local metatable = table.clone(KinematicStateMT)
	metatable.__index = State.new(relativeTo)

	setmetatable(self, metatable)

	return self
end


--[=[
	Creates a new KinematicState instance.
]=]
function KinematicState.newFromKinematicState(kinematicState: KinematicState, relativeTo: KinematicState?): KinematicState
	return KinematicState.new(kinematicState.position, kinematicState.velocity, relativeTo)
end

-- Methods

function KinematicState:getPosition(): Type.Vector3D
	return self.position
end

function KinematicState:getVelocity(): Type.Vector3D
	return self.velocity
end

function KinematicState:getAbsolutePosition(): Type.Vector3D
	if self:hasRelative() then
		return self.position + self:getRelative():getAbsolutePosition()
	else
		return self.position
	end
end

function KinematicState:getAbsoluteVelocity(): Type.Vector3D
	if self:hasRelative() then
		return self.velocity + self:getRelative():getAbsoluteVelocity()
	else
		return self.velocity
	end
end

--[=[
	Applies an acceleration to return a new KinematicState, adjusted for delta.

	@param accleration The acceleration to be applied to the new KinematicState.
	@param delta The time (in seconds) across which the acceleration will be applied.

	@return The new KinematicState.
]=]
function KinematicState:addAcceleration(acceleration: Type.AccelerationState, delta: number?): KinematicState
	local newVelocity: Type.Vector3D = self.velocity + acceleration:getAccelerationVector(delta)
	local newPosition: Type.Vector3D = self.position + newVelocity * (delta or acceleration:getDelta())
	return KinematicState.new(newPosition, newVelocity)
end

--- @Override
function KinematicState:getAbsolute(): KinematicState
	return KinematicState.new(
		self:getAbsolutePosition(),
		self:getAbsoluteVelocity()
	)
end

--- @Override
function KinematicState:consolidateOnce(): KinematicState
	assert(self:hasRelative(), "consolidateOnce() cannot be called on a KinematicState with no RelativeTo")
	local relativeTo: KinematicState = self:getRelative()

 	return KinematicState.new(
			self:getPosition() + relativeTo:getPosition(),
			self:getVelocity() + relativeTo:getVelocity(),
			relativeTo:getRelativeOrNil()
		)
end

--- @Override
function KinematicState:synchronize(other: KinematicState): (KinematicState, KinematicState)
	local selfTree: { KinematicState } = self:getRelativeTree()
	local selfTrimmedPosition: Type.Vector3D = Vector3D.zero
	local selfTrimmedVelocity: Type.Vector3D = Vector3D.zero

	for i = 1, self:convergenceIndex(other) - 1 do
		selfTrimmedPosition += selfTree[i]:getPosition()
		selfTrimmedVelocity += selfTree[i]:getVelocity()
	end

	local otherTree: { KinematicState } = other:getRelativeTree()
	local otherTrimmedPosition: Type.Vector3D = Vector3D.zero
	local otherTrimmedVelocity: Type.Vector3D = Vector3D.zero

	for i = 1, other:convergenceIndex(self) - 1 do
		otherTrimmedPosition += otherTree[i]:getPosition()
		otherTrimmedVelocity += otherTree[i]:getVelocity()
	end

	local convergenceItem: KinematicState = self:convergenceItem(other) or KinematicState.new(Vector3D.zero, Vector3D.zero)

	local selfResult: KinematicState = KinematicState.new(
			convergenceItem.position + selfTrimmedPosition,
			convergenceItem.velocity + selfTrimmedVelocity
		)
	local otherResult: KinematicState = KinematicState.new(
			convergenceItem.position + otherTrimmedPosition,
			convergenceItem.velocity + otherTrimmedVelocity
		)

	assert(
		selfResult:getRelativeOrNil() == otherResult:getRelativeOrNil()
		and self:getAbsolutePosition() == selfResult:getAbsolutePosition()
		and self:getAbsoluteVelocity() == selfResult:getAbsoluteVelocity()
		and other:getAbsolutePosition() == otherResult:getAbsolutePosition()
		and other:getAbsoluteVelocity() == otherResult:getAbsoluteVelocity(),
		"something wrong in the calcs!"
	)

	return selfResult, otherResult
end

--- @Override
function KinematicState:matchRelative(other: KinematicState): KinematicState
	local otherTree: { KinematicState } = other:getRelativeTree()
	local otherTrimmedPosition: Type.Vector3D = Vector3D.zero
	local otherTrimmedVelocity: Type.Vector3D = Vector3D.zero

	-- consolidate other to match with self's RelativeTo tree, and track trimmed stats
	for i = 1, other:convergenceIndex(self) - 1 do
		otherTrimmedPosition += otherTree[i]:getPosition()
		otherTrimmedVelocity += otherTree[i]:getVelocity()
	end

	-- subtract the stats exclusively between other and self, and add the excess to the newly matched result
	local convergenceItem: KinematicState = self:convergenceItem(other) or KinematicState.new(Vector3D.zero, Vector3D.zero)
	local selfRelativeTree: { KinematicState } = self:getRelativeTree()
	local resultPositionLeftover: Type.Vector3D = convergenceItem.position + otherTrimmedPosition
	local resultVelocityLeftover: Type.Vector3D = convergenceItem.velocity + otherTrimmedVelocity

	for i = self:convergenceIndex(other) - 1, 2, -1 do
		resultPositionLeftover -= selfRelativeTree[i].position
		resultVelocityLeftover -= selfRelativeTree[i].velocity
	end

	local result: KinematicState = KinematicState.new(resultPositionLeftover, resultVelocityLeftover, self:getRelativeOrNil())

	assert(
		self:getRelativeOrNil() == result:getRelativeOrNil()
		and other:getAbsolutePosition() == result:getAbsolutePosition()
		and other:getAbsoluteVelocity() == result:getAbsoluteVelocity(),
		"something wrong in the calcs!"
	)

	return result
end

return (KinematicState :: any) :: Constructor.KinematicState

File: src/shared/Modules/BaseModule/Relative/State/KinematicState.ts
// import { $assert } from "rbxts-transform-debug";
import Vector3D from "shared/Modules/Libraries/Vector3D";

import State from ".";
import AccelerationState from "./AccelerationState";

export default class KinematicState extends State {
	public readonly position: Vector3D;
	public readonly velocity: Vector3D;

	// Constructors

	/**
	 * Creates a new KinematicState instance.
	 * @param position 
	 * @param velocity 
	 * @param relativeTo 
	 */
	public constructor(position: Vector3D, velocity: Vector3D, relativeTo?: KinematicState)

	/**
	 * Clones a KinematicState instance and replaces its relativeTo.
	 * @param kinematicState 
	 * @param relativeTo 
	 */
	public constructor(kinematicState: KinematicState, relativeTo?: KinematicState)

	public constructor(arg1: Vector3D | KinematicState, arg2?: Vector3D | KinematicState, arg3?: KinematicState) {
		let position: Vector3D;
		let velocity: Vector3D;
		let relativeTo: KinematicState | undefined

		if (arg1 instanceof Vector3D) { // Constructor 1
			assert(arg2 instanceof Vector3D);
			position = arg1;
			velocity = arg2;
			relativeTo = arg3;
		} else { // Constructor 2
			assert((arg2 === undefined || arg2 instanceof KinematicState) && arg3 === undefined);
			position = arg1.position;
			velocity = arg1.velocity;
			relativeTo = arg2;
		}

		// assert((position.X === position.X) && (position.Y === position.Y) && (position.Z === position.Z), `KinematicState constructor() position is NaN (${position})`);
		// assert((velocity.X === velocity.X) && (velocity.Y === velocity.Y) && (velocity.Z === velocity.Z), `KinematicState constructor() velocity is NaN (${velocity})`);
		super(relativeTo);
		this.position = position;
		this.velocity = velocity;
	}

	// Arithmetic

	public add(other: KinematicState): KinematicState {
		assert(this.sameRelativeTree(other), "KinematicState.add() operands do not share the same relativeTree");
		return new KinematicState(
			this.position.add(other.position),
			this.velocity.add(other.velocity),
			this.getRelativeOrUndefined(),
		);
	}

	public sub(other: KinematicState): KinematicState {
		assert(this.sameRelativeTree(other), "KinematicState.sub() operands do not share the same relativeTree");
		return new KinematicState(
			this.position.sub(other.position),
			this.velocity.sub(other.velocity),
			this.getRelativeOrUndefined(),
		);
	}

	// Comparisons

	public equals(other?: KinematicState): other is KinematicState {
		if (this !== undefined && other !== undefined) {
			if (this.position.equals(other.position) && this.velocity.equals(other.velocity))
				if (this.hasRelative() && other.hasRelative())
					return this.getRelative().equals(other.getRelative());
				else
					return this.hasRelative() === other.hasRelative();
			else
				return false;
		} else {
			return this === undefined && other === undefined;
		}
	}

	// Methods

	public getAbsolutePosition(): Vector3D {
		if (this.hasRelative()) {
			return this.position.add(this.getRelative().getAbsolutePosition());
		} else {
			return this.position;
		}
	}

	public getAbsoluteVelocity(): Vector3D {
		if (this.hasRelative()) {
			return this.velocity.add(this.getRelative().getAbsoluteVelocity());
		} else {
			return this.velocity;
		}
	}

	/**
	 * Advances this KinematicState in time, and recursively does the
	 * same (without acceleration) to its entire relative tree.
	 * @param delta The time to advance.
	 * @param acceleration Optionally applies acceleration over delta.
	 * @returns A new AccelerationState, with all of its relatives also changed.
	 */
	public step(delta: number, acceleration?: AccelerationState): KinematicState {
		const newVelocity: Vector3D = this.velocity;
		const newPosition: Vector3D = this.position;

		if (acceleration !== undefined)
			newVelocity.add(acceleration.getAccelerationVector(delta));
		newPosition.add(newVelocity.mul(delta));

		return new KinematicState(
			newPosition,
			newVelocity,
			this.getRelative().step(delta)
		);
	}

	public getAbsolute(): KinematicState {
		return new KinematicState(this.getAbsolutePosition(), this.getAbsoluteVelocity());
	}

	public consolidateOnce(): KinematicState {
		assert(this.hasRelative(), "consolidateOnce() cannot be called on a KinematicState with no relativeTo");
		const relativeTo = this.getRelative();
		return new KinematicState(
			this.position.add(relativeTo.position),
			this.velocity.add(relativeTo.velocity),
			relativeTo.getRelativeOrUndefined(),
		);
	}

	public synchronize(other: KinematicState): [KinematicState, KinematicState] {
		const selfTree = this.getRelativeTree();
		let selfTrimmedPosition = Vector3D.zero;
		let selfTrimmedVelocity = Vector3D.zero;

		for (let i = 0; i < this.convergenceIndex(other) - 1; i++) {
			selfTrimmedPosition = selfTrimmedPosition.add(selfTree[i].position);
			selfTrimmedVelocity = selfTrimmedVelocity.add(selfTree[i].velocity);
		}

		const otherTree = other.getRelativeTree();
		let otherTrimmedPosition = Vector3D.zero;
		let otherTrimmedVelocity = Vector3D.zero;

		for (let i = 0; i < other.convergenceIndex(this) - 1; i++) {
			otherTrimmedPosition = otherTrimmedPosition.add(otherTree[i].position);
			otherTrimmedVelocity = otherTrimmedVelocity.add(otherTree[i].velocity);
		}

		const convergenceItem = this.convergenceItem(other) ?? new KinematicState(Vector3D.zero, Vector3D.zero);

		const selfResult = new KinematicState(
			convergenceItem.position.add(selfTrimmedPosition),
			convergenceItem.velocity.add(selfTrimmedVelocity),
		);
		const otherResult = new KinematicState(
			convergenceItem.position.add(otherTrimmedPosition),
			convergenceItem.velocity.add(otherTrimmedVelocity),
		);

		assert(
			selfResult.getRelativeOrUndefined() === otherResult.getRelativeOrUndefined() &&
				this.getAbsolutePosition().equals(selfResult.getAbsolutePosition()) &&
				this.getAbsoluteVelocity().equals(selfResult.getAbsoluteVelocity()) &&
				other.getAbsolutePosition().equals(otherResult.getAbsolutePosition()) &&
				other.getAbsoluteVelocity().equals(otherResult.getAbsoluteVelocity()),
			"something wrong in the calcs!",
		);

		return [selfResult, otherResult];
	}

	public matchRelative(other: KinematicState): KinematicState {
		const otherTree = other.getRelativeTree();
		let otherTrimmedPosition = Vector3D.zero;
		let otherTrimmedVelocity = Vector3D.zero;

		for (let i = 0; i < other.convergenceIndex(this) - 1; i++) {
			otherTrimmedPosition = otherTrimmedPosition.add(otherTree[i].position);
			otherTrimmedVelocity = otherTrimmedVelocity.add(otherTree[i].velocity);
		}

		const convergenceItem = this.convergenceItem(other) ?? new KinematicState(Vector3D.zero, Vector3D.zero);
		const selfRelativeTree = this.getRelativeTree();

		let resultPositionLeftover = convergenceItem.position.add(otherTrimmedPosition);
		let resultVelocityLeftover = convergenceItem.velocity.add(otherTrimmedVelocity);

		for (let i = this.convergenceIndex(other) - 1; i >= 1; i--) {
			resultPositionLeftover = resultPositionLeftover.sub(selfRelativeTree[i].position);
			resultVelocityLeftover = resultVelocityLeftover.sub(selfRelativeTree[i].velocity);
		}

		const result = new KinematicState(resultPositionLeftover, resultVelocityLeftover, this.getRelativeOrUndefined());

		assert(
			this.getRelativeOrUndefined() === result.getRelativeOrUndefined() &&
				other.getAbsolutePosition().equals(result.getAbsolutePosition()) &&
				other.getAbsoluteVelocity().equals(result.getAbsoluteVelocity()),
			"something wrong in the calcs!",
		);

		return result;
	}

	// Wrap super methods with current type

	override convergenceIndex(other: KinematicState): number {
		return super.convergenceIndex(other);
	}

	override getRelative(): KinematicState {
		return super.getRelative() as KinematicState;
	}

	override getRelativeOrUndefined() : KinematicState | undefined {
		return super.getRelativeOrUndefined() as KinematicState | undefined;
	}

	override getRelativeTree(): KinematicState[] {
		return super.getRelativeTree() as KinematicState[];
	}

	override convergenceItem(other: KinematicState): KinematicState | undefined {
		return super.convergenceItem(other) as KinematicState | undefined;
	}

	override deepClone(): KinematicState {
		return this;
	}
}
